packages/flutter/lib/src/material/icon_button_theme.dart

```diff
@@ -0,0 +1,123 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'button_style.dart';
+import 'theme.dart';
+
+/// A [ButtonStyle] that overrides the default appearance of
+/// [IconButton]s when it's used with [IconButtonTheme] or with the
+/// overall [Theme]'s [ThemeData.iconButtonTheme].
+///
+/// The [IconButton] will be affected by [IconButtonTheme] and [IconButtonThemeData]
+/// only if [ThemeData.useMaterial3] is set to true; otherwise, [IconTheme] will be used.
```

Comment:
If we don't factor in the icon theme when useMaterial3 is true, then this is a significant breaking change. If the IconButtonTheme overrides IconTheme, then we might avoid a breaking change here, however detecting the difference between an intentional IconTheme override and the default IconTheme created by the Theme widget will be a little bit tricky.

Reply:
I think we can avoid making a breaking change, details are here: [IconTheme and IconButton defaults](https://docs.google.com/document/d/1vnGo4WzW_7JSJS8nFs2GNkaEZ8cx6T-odBeg7du0I_M/edit?usp=sharing
). The gist of the proposal is this:

IconButton’s default size and the color of its Icon descendant are currently defined by the IconTheme. We’d like them to be defined by the new IconButtonTheme instead, if an IconButtonTheme is defined with non-null properties. Otherwise we’d like them to be defined by the defaults for Material 3 rather than the current IconTheme.  At least not by the current IconTheme, _if_ the current IconTheme is the default one automatically defined by Theme.

We could check for that case:
```dart
  IconThemeData overallIconTheme = IconTheme.of(context);
  IconThemeData? iconTheme = isThemeIconTheme(overallIconTheme)
    ? null 
    : overallIconTheme;

  bool isThemeIconTheme(IconThemeData theme) {
     // Return true if theme is the same as the default 
     // value of Theme.of(context).iconTheme
  }
```
In this case iconTheme would only be used to define the icon button’s default size and color if it was non-null, i.e. if it wasn’t the default one defined by the overall Theme. That means that if the app has inserted its own IconTheme into the widget tree, we’d use that instead of the M3 defaults. Just to clarify, here’s how we’d resolve each IconButton property, in order of precedence:

1. Widget properties like IconButton.color
2. IconButtonTheme properties
3. IconTheme properties unless they were defined by Theme.of(context).iconTheme
4. Material 3 default IconButtonTheme properties

As always, the first non-null value is the one that’s used. As noted in the code sample, the exception for IconTheme properties is for ones defined by the default value of ThemeData.iconTheme.

Reply:
CC @rydmike - thought you might be interested in this issue since it's related to the one that came up in https://github.com/flutter/flutter/issues/107305#issuecomment-1181680901

Reply:
Thanks @HansMuller, I like this approach. It will work without breaking past M2 behavior and work as one might expect when using M3 and the new `IconButtonTheme` as well.

Reply:
Thanks a lot for the suggestions. I just pushed some changes based on the comments so both IconTheme and IconButtonTheme should work in M3 now. Please let me know if there're any questions:)